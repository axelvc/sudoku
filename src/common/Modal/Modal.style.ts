import styled from 'styled-components'

export type variant = 'error'

export const Container = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  z-index: 10;
  display: grid;
  place-items: center;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.7);
`

export const StyledModal = styled.section`
  position: relative;
  display: flex;
  flex-direction: column;
  gap: 2rem;
  width: fill-available;
  max-width: 26rem;
  padding: 2rem;
  margin: 1.5rem;
  overflow: hidden;
  text-align: center;
  background: var(--color-gray-100);
  border-radius: 1rem;
`

interface TitleProps {
  variant?: variant
}

export const Title = styled.h2<TitleProps>`
  font-size: 1.5rem;
  font-weight: 700;
  color: ${p => p.variant === 'error' && 'var(--color-red-600);'};
`
