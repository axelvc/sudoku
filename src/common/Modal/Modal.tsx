import { FC } from 'react'
import { createPortal } from 'react-dom'

import { Container, StyledModal, Title, variant } from './Modal.style'

interface Props {
  title: string
  variant?: variant
}

const Modal: FC<Props> = ({ title, children, variant }) => {
  return createPortal(
    <Container data-testid="modal-container">
      <StyledModal data-testid="modal">
        <Title variant={variant}>{title}</Title>
        {children}
      </StyledModal>
    </Container>,
    document.body,
  )
}

export default Modal
