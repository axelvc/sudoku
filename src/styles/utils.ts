import { css } from 'styled-components'

export const breakpoints = {
  md: '768px',
}

export const roundedGrid = (rows: number, cols: number): ReturnType<typeof css> => css`
  display: grid;
  grid-template: ${p => `repeat(${rows}, 1fr) / repeat(${cols}, 1fr) `};
  gap: 0.25rem;
  overflow: hidden;
  border-radius: 0.5rem;
`

export const fancyBox = css`
  padding: 1.5rem;
  background: var(--color-gray-100);
  border-radius: 1rem;
`
