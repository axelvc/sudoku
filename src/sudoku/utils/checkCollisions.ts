import { Coords, BoxData, SudokuState } from '../sudokuSlice'

function getRowSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  return puzzle[row].filter((box, i) => box.value && i !== col)
}

function getColSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  return puzzle.reduce((siblings, r, i) => {
    const box = r[col]

    if (box.value && i !== row) {
      siblings.push(box)
    }

    return siblings
  }, [])
}

function getBlockSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  const blockRow = Math.floor(row / 3)
  const blockCol = Math.floor(col / 3)

  return puzzle.reduce((siblings, r, rI) => {
    r.forEach((box, cI) => {
      if (
        box.value &&
        rI !== row &&
        cI !== col &&
        blockRow === Math.floor(rI / 3) &&
        blockCol === Math.floor(cI / 3)
      ) {
        siblings.push(box)
      }
    })

    return siblings
  }, [])
}

export default function checkCollisions(
  state: SudokuState,
  { row, col }: Coords,
  oldValue: number,
): void {
  const { puzzle } = state
  const box = puzzle[row][col]
  const siblings = [
    ...getRowSiblings(puzzle, { row, col }),
    ...getColSiblings(puzzle, { row, col }),
    ...getBlockSiblings(puzzle, { row, col }),
  ]

  if (box.errors) {
    box.errors = 0

    siblings.forEach(sibling => {
      if (sibling.value === oldValue) {
        sibling.errors--
      }
    })
  }

  siblings.forEach(sibling => {
    if (sibling.value === box.value) {
      sibling.errors++
      box.errors++
    }
  })
}
