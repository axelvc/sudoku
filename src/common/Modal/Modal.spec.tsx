import { regex, render, screen } from '../../test/utils'

import Modal from './Modal'

describe('Modal component', () => {
  it('should render correct title and details', () => {
    const title = 'A sample title to test'
    const details =
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Consectetur, illum.'

    render(
      <Modal title={title}>
        <p>{details}</p>
        <button>ok</button>
      </Modal>,
    )

    const $modal = screen.getByTestId('modal-container')

    expect($modal).toHaveTextContent(regex(title))
    expect($modal).toHaveTextContent(regex(details))
    expect($modal).toContainHTML('<button>ok</button>')
  })

  it('should have same styles', () => {
    render(<Modal title="" />)

    const $modal = screen.getByTestId('modal-container')

    expect($modal).toMatchSnapshot()
  })

  it('should have error title styles', () => {
    render(<Modal title="" variant="error" />)

    const $modal = screen.getByTestId('modal-container')

    expect($modal).toMatchInlineSnapshot(`
.c0 {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 10;
  display: grid;
  place-items: center;
  width: 100%;
  height: 100%;
  background: rgba(0,0,0,0.7);
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 2rem;
  max-width: 26rem;
  padding: 2rem;
  margin: 1.5rem;
  overflow: hidden;
  text-align: center;
  background: var(--color-gray-100);
  border-radius: 1rem;
}

.c2 {
  font-size: 1.5rem;
  font-weight: 700;
  color: var(--color-red-600);
}

<div
  class="c0"
  data-testid="modal-container"
>
  <section
    class="c1"
    data-testid="modal"
  >
    <h2
      class="c2"
    />
  </section>
</div>
`)
  })
})
