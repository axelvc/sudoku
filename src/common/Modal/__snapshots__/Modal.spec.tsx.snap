// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should have same styles 1`] = `
.c0 {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 10;
  display: grid;
  place-items: center;
  width: 100%;
  height: 100%;
  background: rgba(0,0,0,0.7);
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 2rem;
  max-width: 26rem;
  padding: 2rem;
  margin: 1.5rem;
  overflow: hidden;
  text-align: center;
  background: var(--color-gray-100);
  border-radius: 1rem;
}

.c2 {
  font-size: 1.5rem;
  font-weight: 700;
}

<div
  class="c0"
  data-testid="modal-container"
>
  <section
    class="c1"
    data-testid="modal"
  >
    <h2
      class="c2"
    />
  </section>
</div>
`;
